clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -1038,16 +1051,30 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
                 requestRejoin();
                 future.raise(error);
             } else if (error == Errors.ILLEGAL_GENERATION) {
-                log.info("Attempt to heartbeat failed since generation {} is not current", generation.generationId);
-                resetGenerationOnResponseError(ApiKeys.HEARTBEAT, error);
-                future.raise(error);
+                if (sentGeneration.equals(currentGeneration)) {
```

Comment:
This is basically the same thing that we have below. I think we could move it into a generic function in `CoordinatorResponseHandler` so that it could be used by other handlers.